google/cloud/storage/oauth2/credentials.h

```diff
@@ -41,6 +41,22 @@ class Credentials {
    */
   virtual std::pair<google::cloud::storage::Status, std::string>
   AuthorizationHeader() = 0;
+
+  /**
+   * Sign a blob using the credentials.
+   *
+   * Create a RSA SHA256 signature of the blob using the Credential object. If
+   * the credentials do not support signing blobs it returns an error status.
+   *
+   * @param blob the bytes to sign.
+   * @return a Base64-encoded RSA SHA256 digest of @p blob using the current
+   *   credentials.
+   */
+  virtual std::pair<google::cloud::storage::Status, std::string> SignBlob(
```

Comment:
I don't think this belongs on the `Credentials` interface class.  Would it be cleaner to check the type of the `Credentials` object at runtime, making sure that it's been upcasted from a `ServiceAccountCredentials`, and failing if not?  That would avoid polluting the other subclasses with these method overrides that just return an error.

For checking the "real" type of a credential subclass that's been upcasted to a `Credentials`, we could either give each subclass a constant type attribute to check, or maybe perform some `dynamic_cast` logic to check the type (not sure how safe this is, haven't done the reading yet).  Thoughts?

Reply:
`dynamic_cast<>` is perfectly safe (for pointers, it returns nullptr on failure, for references it raises), let's do that.